Implement SystemcoreClock

TomoYamanaka opened the pull request:
Hi.

We implemented SystemcoreClock which is defined in CMSIS.

Regards;
Yamanaka

0xc0170 commented:
Does this mean that clock is fixed and can't be changed? 

A user case:
```
extern uint32_t SystemCoreClock;

printf("Old core clock: %u \n", SystemCoreClock);
user_change_clocks();
SystemCoreClockUpdate();
printf("New core clock: %u \n", SystemCoreClock);
```

TomoYamanaka commented:
Hi

As shown below, other companies also implement by the same method.
Do you have any problems?
- \hal\targets\cmsis\TARGET_ARM_SSG\TARGET_IOTSS_BEID\system_CMSDK_BEID.c
- \hal\targets\cmsis\TARGET_Atmel\TARGET_SAM_CortexM0P\utils\cmsis\TARGET_SAMD21\
   source\system_samd21.c

and more...

We think that SystemCoreClock is to store the CPU Clock.

0xc0170 commented:
Good point, they make it wrong ! That does not update the core clock variable. The docs defines it clearly: https://www.keil.com/pack/doc/CMSIS/Core/html/group__system__init__gr.html#gae0c36a9591fe6e9c45ecb21a794f0f0f

we should review those CoreClockUpdate functions for targets, identify those that don't do it properly.

0xc0170 commented:
I created an issue for the rest of the targets, can you please update the Update function to read the value from registers?

TomoYamanaka commented:
We will consider about the implement method of Update function.

TomoYamanaka commented:
Hi.

We implemented SystemCoreClockUpdate function.
In SystemCoreClockUpdate function, we changed to calculare the CPU Clock by the division ratio setting of from FRQCR register.

Regards;
Yamanaka

0xc0170 commented:
Thank you for the update
